source/Plugins.cpp

```diff
@@ -345,6 +507,132 @@ const Set<Plugin> &Plugins::Get()
 // Toggles enabling or disabling a plugin for the next game restart.
 void Plugins::TogglePlugin(const string &name)
 {
+	lock_guard<mutex> guard(pluginsMutex);
 	auto *plugin = plugins.Get(name);
 	plugin->currentState = !plugin->currentState;
 }
+
+
+
+future<void> Plugins::Install(InstallData *installData, bool update)
+{
+	oldNetworkActivity = true;
+	if(!update)
+	{
+		lock_guard<mutex> guard(activePluginsMutex);
+		if(!activePlugins.insert(installData->name).second)
+			return future<void>();
+	}
+
+	return async(launch::async, [installData, update]() noexcept -> void
+		{
+			string zipLocation = Files::Plugins() + installData->name + ".zip";
+			bool success = Download(installData->url, zipLocation);
+			if(success)
+			{
+				success = ExtractZIP(
+					zipLocation,
+					Files::Plugins(), installData->name + "/");
+				if(success)
+				{
+					if(update)
+						Files::DeleteDir(Files::Plugins() + installData->name);
+					// Create a new entry for the plugin.
+					Plugin *newPlugin;
+					{
+						lock_guard<mutex> guard(pluginsMutex);
+						newPlugin = plugins.Get(installData->name);
+					}
+					newPlugin->name = installData->name;
+					newPlugin->aboutText = installData->aboutText;
+					newPlugin->path = Files::Plugins() + installData->name + "/";
+					newPlugin->version = installData->version;
+					newPlugin->enabled = false;
+					newPlugin->currentState = true;
+
+					installData->installed = true;
+					installData->outdated = false;
+				}
+				else
+					Files::DeleteDir(Files::Plugins() + installData->name);
+			}
+			Files::Delete(zipLocation);
+			{
+				lock_guard<mutex> guard(activePluginsMutex);
+				activePlugins.erase(installData->name);
+			}
+		});
+}
+
+
+
+future<void> Plugins::Update(InstallData *installData)
+{
+	{
+		lock_guard<mutex> guard(activePluginsMutex);
+		if(!activePlugins.insert(installData->name).second)
+			return future<void>();
+	}
+
+	{
+		lock_guard<mutex> guard(pluginsMutex);
+		plugins.Get(installData->name)->version = installData->version;
+	}
+
+	return Install(installData, true);
```

Comment:
Install doesn't delete the old version of the plugin, does it?

Reply:
it does